slides/install-rspec.markdown

```diff
@@ -0,0 +1,305 @@
+
+---
+
+## RSpec best practice
+
+Best practice include ideas how to improve your specs quality and increase efficiency of your BDD workflow.
+
+--
+
+### `describe`
+
+Be clear about what method you're describing. For instance, use the ruby documentation convention of `.` when referring to a class method's name and `#` when referring to an instance method's name.
+
+```ruby
+### Bad example
+describe 'the authenticate method for User' do
+describe 'if the user is an admin' do
+
+### Good example
+describe '.authenticate' do
+describe '#admin?' do
+```
+
+--
+
+### `context`
+
+`context` starts either with `"with"` or `"when"`, such "when status is pending"
+
+```ruby
+describe '#status_badge' do
+  context 'returns css class based on status' do
+    context 'when status is pending' do
+      let(:request_status) { 'pending' }
+
+      it 'returns css for grey badge' do
+        expect(subject.status_badge).to eql 'c-badge--grey'
+      end
+    end
+  end
+end
+```
+
+--
+
+### `it`
+`it` describes a test case (one expectation) and specify only one behavior. Multiple expectations in the same example are signal that you may be specifying multiple behaviors. By specify only have one expectation, helps you on finding possible errors, going directly to the failing test, and to make your code readable.
+
+```ruby
+it { is_expected.to belong_to(:job_title).class_name('JobTitle').optional }
+```
+
+or
+
+```ruby
+it 'has relations' do
+  is_expected.to belong_to(:job_title).class_name('JobTitle').optional
+end
+```
+
+--
+
+### `subject`
+
+`subject` makes way clearer about what you're actually testing and helps you stay DRY in your tests
+
+```ruby
+describe Book do
+  describe '#valid_isbn?' do
+    subject { Book.new(isbn: isbn).valid_isbn? }
+
+    context 'with a valid ISBN number' do
+      let(:isbn) { 'valid' }
+
+      # ...
+    end
+  end
+end
+```
+
+--
+
+### `Mocking`
+
+`mocking` is interesting and usually we're doing mocking when the scenario which we want to test require another service.
+
+You may mock just everything so your spec will never hit the database or another service. But, this is something wrong. When your model code changed or the initiliaze method of service you are call changed, your code will break without get failing specs before merge to production.
+
+app/models/post_info.rb <!-- .element: class="filename" -->
+
+```ruby
+class PostInfo do
+  #...
+  def allow?
+    current_company.active_subscribe?
+  end
+end
+```
+
+spec/models/post_info_spec.rb <!-- .element: class="filename" -->
+
+```ruby
+RSpec.describe PostInfo do
+  before do
+    allow_any_instance_of(Company).to receive(:active_subscribe?).and_return true
+  end
+
+  it ... do
+  #..
+end
+```
+
+--
+
+## Custom matchers
+
+Rspec has many useful matchers, we already used be true, be false, and so on. Sometimes, when expecting given values, we repeat the same code over and over again.
+
+```ruby
+RSpec.feature 'some feature', type: :feature do
```

Comment:
`RSpec.feature 'some feature', type: :feature do`, also the same is `feature 'some feature' do`, read about it